packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -0,0 +1,352 @@
+            logWarn(
+                'Hydration mismatch: comment values do not match, will recover from the difference',
+                vnode.owner
+            );
+        }
+    }
+
+    // always set the text value to the one from the vnode.
+    node.nodeValue = vnode.text ?? null;
+    vnode.elm = node;
+}
+
+function hydrateElement(vnode: VElement, node: Node) {
+    if (process.env.NODE_ENV !== 'production') {
+        // eslint-disable-next-line lwc-internal/no-global-node
+        validateNodeType<Element>(vnode, node, Node.ELEMENT_NODE);
+        validateElement(vnode, node);
+    }
+
+    const elm = node as Element;
+    vnode.elm = elm;
+
+    const { context } = vnode.data;
+    const isDomManual = Boolean(
+        !isUndefined(context) && !isUndefined(context.lwc) && context.lwc.dom === LwcDomMode.Manual
+    );
+
+    if (isDomManual) {
+        // it may be that this element has lwc:inner-html, we need to diff and in case are the same,
+        // remove the innerHTML from props so it reuses the existing dom elements.
+        const { props } = vnode.data;
+        if (!isUndefined(props) && !isUndefined(props.innerHTML)) {
+            if (elm.innerHTML === props.innerHTML) {
+                delete props.innerHTML;
+            } else {
+                logWarn(
+                    `Mismatch hydrating element <${elm.tagName.toLowerCase()}>: innerHTML values do not match for element, will recover from the difference`,
+                    vnode.owner
+                );
+            }
+        }
+    }
+
+    patchElementPropsAndAttrs(vnode);
+
+    if (!isDomManual) {
+        hydrateChildren(vnode.elm.childNodes, vnode.children, vnode.owner);
+    }
+}
+
+function hydrateCustomElement(vnode: VCustomElement, node: Node) {
+    if (process.env.NODE_ENV !== 'production') {
+        // eslint-disable-next-line lwc-internal/no-global-node
+        validateNodeType<Element>(vnode, node, Node.ELEMENT_NODE);
+        validateElement(vnode, node);
+    }
+
+    const elm = node as Element;
+    vnode.elm = elm;
+
+    const { sel, mode, ctor, owner } = vnode;
+
+    const def = getComponentInternalDef(ctor);
+    createVM(elm, def, {
+        mode,
+        owner,
+        tagName: sel,
+    });
+
+    const vm = getAssociatedVM(elm);
+    allocateChildren(vnode, vm);
+
+    patchElementPropsAndAttrs(vnode);
+
+    // Insert hook section:
+    if (process.env.NODE_ENV !== 'production') {
+        assert.isTrue(vm.state === VMState.created, `${vm} cannot be recycled.`);
+    }
+    runConnectedCallback(vm);
+
+    if (vm.renderMode !== RenderMode.Light) {
+        // VM is not rendering in Light DOM, we can proceed and hydrate the slotted content.
+        // Note: for Light DOM, this is handled while hydrating the VM
+        hydrateChildren(vnode.elm.childNodes, vnode.children, vm);
+    }
+
+    hydrateVM(vm);
+}
+
+export function hydrateChildren(elmChildren: NodeListOf<ChildNode>, children: VNodes, vm: VM) {
```

Comment:
This seems to be the main entry point into the hydration logic. We don't have to solve this in this PR, but I think it would be worth thinking about how we could tree-shake this function for environments that don't use hydration. (Now that it's actually possible.)

E.g. Svelte has a [`hydratable`](https://svelte.dev/docs#compile-time-svelte-compile) option which, when set to false in the compiler, can remove most of the hydration logic from the bundle output. For our case, we may also be able to use an environment variable or something.

Reply:
> This seems to be the main entry point into the hydration logic. We don't have to solve this in this PR, but I think it would be worth thinking about how we could tree-shake this function for environments that don't use hydration. (Now that it's actually possible.)

Tree-shaking wasn't possible prior to this the hydration logic was coupled with the rendering logic. After this refactor I can confirm that the rehydration logic is completely tree-shaken if not imported.